guix: Don't include directory name in SHA256SUMS

achow101 opened the pull request:
The SHA256SUMS file can be used in a sha256sum -c command to verify downloaded binaries. However users are likely to download just a single file and not place this file in the correct directory relative to the SHA256SUMS file for the simple verification command to work. By not including the directory name in the SHA256SUMS file, it will be easier for users to verify downloaded binaries.

achow101 committed:
guix: Don't include directory name in SHA256SUMS

The SHA256SUMS file can be used in a sha256sum -c command to verify
downloaded binaries. However users are likely to download just a single
file and not place this file in the correct directory relative to the
SHA256SUMS file for the simple verification command to work. By not
including the directory name in the SHA256SUMS file, it will be easier
for users to verify downloaded binaries.

Co-authored-by: Carl Dong <contact@carldong.me>

wodry commented on contrib/guix/libexec/build.sh:
```diff
@@ -452,11 +452,7 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
-        find "$ACTUAL_OUTDIR" -type f
-    } | xargs realpath --relative-base="$PWD" \
-      | xargs sha256sum \
+    find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$1" | xargs sha256sum' _ {} \; \
```
To make find print the filename without path, you can simply use `find -printf "%f\n"`, this would result in:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -printf "%f\n" | xargs sha256sum' _ {}`

wodry commented:
Why don't you mention this PR in https://github.com/bitcoin/bitcoin/issues/22634#issuecomment-894463139 ?

I have now rewritten my download&verfiy bash script already. It's true that I e.g. do not download into a subfolder "x86_64-linux-gnu", so as you said, I experienced that sha256sum -c --ignore-missing does not work.

My workaround was just to do grep <Tarballname> SHA256SUMS and then sha256sum <Tarballname> to be able to do manual visual comparison of the two output hashes (which I did also with the previous SHA256SUM format file, which contained the signature itself).

I am not sure if this is good or bad, on one hand I like it that the SHA256SUMS file mirrors the download directory structure, that is clean and I think it would not be feasible to have an SHA256SUMS file in every arch sub dir, because that would require also a signature file for each of them. On the other hand, I agree, that normally one does not download in an arch sub dir. But the latter could be done easily I think, though.

hebasto commented:
Concept ACK. Arch is included in the output filenames, therefore, no name clash could be expected.

hebasto commented on contrib/guix/libexec/build.sh:
```diff
@@ -452,11 +452,7 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
-        find "$ACTUAL_OUTDIR" -type f
-    } | xargs realpath --relative-base="$PWD" \
-      | xargs sha256sum \
+    find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$1" | xargs sha256sum' _ {} \; \
```
How does the `_` variable work here?

The Bash [doc](https://www.gnu.org/software/bash/manual/bash.html#Bash-Variables) did not help much.

wodry replied:
LOL I just tried for 30 min to figure out :) I am not bad in Bash, but this I just can not understand.

Zero-1729 replied:
AFAIK, post shell startup, `_` stores the arguments of the last executed command. E.g. it stores `-G` after the `ls -G` command is executed.

So in the case above, it should store the pathname `/outdir-base`, i.e. the argument of the last command executed before this line (`cd /outdir-base`). 

Unless I am missing something here.

wodry replied:
I guess the _ feeds the bash command, so it results in $1 there, but not sure :)

Zero-1729 replied:
I would assume it works something like that, but honestly not sure at this point ¯\\\_(ツ)_/¯.

achow101 replied:
Shellcheck told me to do it this way: https://github.com/koalaman/shellcheck/wiki/SC2156

hebasto replied:
> Shellcheck told me to do it this way: https://github.com/koalaman/shellcheck/wiki/SC2156

TIL, thanks!

wodry replied:
This is really Bash voodoo, I wonder how many people in the world understand this. Please add a comment, e.g. _"The _ is a dummy string that becomes $0 in the bash command. {} becomes $1 in the bash command. Recommended by shellcheck, see https://github.com/koalaman/shellcheck/wiki/SC2156"_ so more people can understand the code.

Zero-1729 replied:
ACK on adding that comment (or anything similar). It would save many hours of trying to grok this.

achow101 replied:
I've added a comment.

wodry replied:
I had still the ambition to understand this strange `_` "dummy" parameter and to find any documentation about that, but as Hebasto said, Bash manpage does not help on `_`. After some testing I think I found out now :)
Manpage of Bash 5.0.3 for option `-c`:

> If the -c option is present, then commands are read from the first non-option argument command_string.
> If there are arguments after the command_string,  the first argument is assigned to $0
> and any remaining arguments are assigned to the positional parameters.
> The assignment to $0 sets the name of the shell, which is used in warning and error messages.

Unfortunately, I do not understand what the last sentence means, can anybody explain? But the text before the last sentence can be proofed with this example:
```
bash -c 'echo $0 $1' param1 param2
param1 param2
```
So what we know from Bash scripts, that $0 is the name of the script, is not the case in bash command using `-c`. In the latter case, $0 is the first parameter, what we know from Bash scripts as $1.
So I think the `_` is only used to fill $0, so the numbering of parameters using bash command -c is the same as in bash script (first really used parameter $1 and so on).
It is not really needed.
You can just replace $1 with $0 in the bash command, then you do not need the `_` dummy parameter.
So you can write shorter and simpler this:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$0" | xargs sha256sum' {} \; \`
Shellcheck is also happy with this.
I would suggest this, as it is simpler and shorter and less confusing. There is only need to explain that with `-c` what one normally knows as $1 becomes $0 there.
A comment to explain the parameter numbering shift is needed anyway I think, be it using the `_` dummy parameter or not. So one can comment like this:

`Unlike in Bash scripts, the first parameter to bash -c <command> is assigned to $0, not $1`

I updated the shellcheck [wiki page](https://github.com/koalaman/shellcheck/wiki/SC2156) to remove the confusing dummy parameter usage.

wodry replied:
I had still the ambition to understand this strange `_` "dummy" parameter and to find any documentation about that, but as Hebasto said, Bash manpage does not help on `_`. After some testing I think I found out now :)
Manpage of Bash 5.0.3 for option `-c`:

> If the -c option is present, then commands are read from the first non-option argument command_string.
> If there are arguments after the command_string,  the first argument is assigned to $0
> and any remaining arguments are assigned to the positional parameters.
> The assignment to $0 sets the name of the shell, which is used in warning and error messages.

Unfortunately, I do not understand what the last sentence means, can anybody explain? But the text before the last sentence can be proofed with this example:
```
bash -c 'echo $0 $1' param1 param2
param1 param2
```
So what we know from Bash scripts, that $0 is the name of the script, is not the case in bash command using `-c`. In the latter case, $0 is the first parameter, what we know from Bash scripts as $1.
So I think the `_` is only used to fill $0, so the numbering of parameters using bash command -c is the same as in bash script (first really used parameter $1 and so on).
It is not really needed.
You can just replace $1 with $0 in the bash command, then you do not need the `_` dummy parameter.
So you can write shorter and simpler this:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$0" | xargs sha256sum' {} \; \`
Shellcheck is also happy with this.
I would suggest this, as it is simpler and shorter and less confusing. There is only need to explain that with `-c` what one normally knows as $1 becomes $0 there.
A comment to explain the parameter numbering shift is needed anyway I think, be it using the `_` dummy parameter or not. So one can comment like this:

`Unlike in Bash scripts, the first parameter to bash -c <command> is assigned to $0, not $1`

I updated the shellcheck [wiki page](https://github.com/koalaman/shellcheck/wiki/SC2156) to remove the confusing dummy parameter usage.

achow101 replied:
I think it's generally discouraged to used `$0` as an actual parameter. Bash's parameter expansion (through `$*` and `$@`) always starts with 1 because 0 is always the filename. If you consider a general case, people may want to use `$*` or `$@` in their commands, in which case there would be an off by one error if this was not mentioned. We should probably have this discussion in the shellcheck repo rather than here.

I'm going to leave this as is since I'm not confident the using `$0` is a good idea.

wodry replied:
OK with dummy parameter then, sounds reasonable, but I would recommend to not name it "_" but "dollarzero" or something, like I have written, to avoid the confusion seen here.

wodry replied:
Just tested and can confirm:
```
bash -c 'echo $*' param1 param2
param2
```

wodry replied:
So i think the confusion for non-Bash-Gurus, while technically being correct, can be boiled down to unluckily, because mistakable, dummy parameter name and insufficient documentation of the reasonability to use it An appropriate update of the shellcheck wiki page would be great.

wodry replied:
> Manpage of Bash 5.0.3 for option `-c`:
>
> > The assignment to $0 sets the name of the shell, which is used in warning and error messages.

I think what they mean is:
"$0 in the shell command is used as the name of the shell, which is used in warning and error messages."

Examples:
```
$ provoke_error
bash: provoke_error: command not found
# -> "bash" is the name of the shell in the error message

$ bash -c 'provoke_error' bash
bash: provoke_error: command not found
# -> "bash" is the name of the shell in the error message

$ bash -c 'provoke_error' mySuperShell
mySuperShell: provoke_error: command not found
# -> "mySuperShell" is the name of the shell in the error message

$ bash -c 'provoke_error' $SHELL
/bin/bash: provoke_error: command not found
# -> "/bin/bash" is the name of the shell in the error message,
#                       which was taken from the command invoking shell's environment variable $SHELL
```

wodry replied:
I have updated https://github.com/koalaman/shellcheck/wiki/SC2156

achow101 replied:
I've changed this to use `bash` instead of `_`.

hebasto reviewed (approved):
ACK 2dabd97f95bec3041b82c21132ce07d76fd1ccca, I have reviewed the code and it looks OK, I agree it can be merged.

Zero-1729 reviewed (commented):
ACK 2dabd97

LGTM, although, I would suggest adding the comment proposed here https://github.com/bitcoin/bitcoin/pull/22654#discussion_r684670218 to make future code review easier.

hebasto commented on contrib/guix/libexec/build.sh:
```diff
@@ -452,11 +452,13 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
-    } | xargs realpath --relative-base="$PWD" \
-      | xargs sha256sum \
+    # Get all of the generated output files and hash them without directory names.
+    # Because "find" prints filenames with a preceeding "./", we need to pass the filenames into "basename" first.
```
typo: preceeding ==> preceding

achow101 replied:
Fixed

hebasto reviewed (approved):
re-ACK 568ef8db0d4cda451b2d324fefad3b854d5ad7b5, only a comment added since my [previous](https://github.com/bitcoin/bitcoin/pull/22654#pullrequestreview-724816208) review.

Zero-1729 reviewed (approved):
re-ACK 568ef8db0d4cda451b2d324fefad3b854d5ad7b5

Nice comment addition.

wodry reviewed (changes requested):
I had still the ambition to understand this strange `_` "dummy" parameter and to find any documentation about that, but as Hebasto said, Bash manpage does not help on `_`. After some testing I think I found out now :)
Manpage of Bash 5.0.3 for option `-c`:

> If the -c option is present, then commands are read from the first non-option argument command_string.
> If there are arguments after the command_string,  the first argument is assigned to $0
> and any remaining arguments are assigned to the positional parameters.
> The assignment to $0 sets the name of the shell, which is used in warning and error messages.

Unfortunately, I do not understand what the last sentence means, can anybody explain? But the text before the last sentence can be proofed with this example:
```
bash -c 'echo $0 $1' param1 param2
param1 param2
```
So what we know from Bash scripts, that $0 is the name of the script, is not the case in bash command using `-c`. In the latter case, $0 is the first parameter, what we know from Bash scripts as $1.
So I think the `_` is only used to fill $0, so the numbering of parameters using bash command -c is the same as in bash script (first really used parameter $1 and so on).
It is not really needed.
You can just replace $1 with $0 in the bash command, then you do not need the `_` dummy parameter.
So you can write shorter and simpler this:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$0" | xargs sha256sum' {} \; \`
Shellcheck is also happy with this.
I would suggest this, as it is simpler and shorter and less confusing. There is only need to explain that with `-c` what one normally knows as $1 becomes $0 there.
A comment to explain the parameter numbering shift is needed anyway I think, be it using the `_` dummy parameter or not. So one can comment like this:

`Unlike in Bash scripts, the first parameter to bash -c <command> is assigned to $0, not $1`

I updated the shellcheck [wiki page](https://github.com/koalaman/shellcheck/wiki/SC2156) to remove the confusing dummy parameter usage.

P.S.
Sorry that my comment is now in the "Conversation" section, which I think is not the correct place for it. I only tried out the "suggest change" button, and now I cannot find out how to remove this and add just a comment in the "Files changed" tab.

Zero-1729 commented:
Great find @wodry! I think as suggested [here](https://github.com/bitcoin/bitcoin/pull/22654/files#r684884983), simplifying the command and amending the comments is probably best. 

Also, this change is now implicitly required as the [shellcheck wiki](https://github.com/koalaman/shellcheck/wiki/SC2156) has been updated to document the new  suggested command kind.

wodry commented:
> Concept ACK. Arch is included in the output filenames, therefore, no name clash could be expected.

Just asking, why arch subfolders at all, if every 22.0rc2 subfolder contains just one file with a unique name which contains the arch, except the folder x86_64-w64-mingw32/ which has two files in it, also without clash?

Zero-1729 commented:
re-ACK cc66a5c27fe029a0e4e52128a567c12d7f900699

laanwj commented:
As I understand this makes the uploaded binaries have all files in one directory like previous releases? (e.g. https://bitcoincore.org/bin/bitcoin-core-0.21.1/ instead of https://bitcoincore.org/bin/bitcoin-core-22.0/test.rc2/)

I didn't think nesting the directory per platform makes anything easier or better, but does add an inconvenient extra click when downloading. So concept ACK.

> I have now rewritten my download&verfiy bash script already.

I would suggest holding off doing any large tool rewrites based on process changes until 22.0 final. Things are still being hammered out.

wodry reviewed (approved):
Code change ACK, Concept 0

wodry commented:
> As I understand this makes the uploaded binaries have all files in one directory like previous releases?

I am not sure what you are referring to with "this"?
This PR leaves the arch sub dir structure as is, only removes the directories from the SHA256SUMS file.
Not sure what and why you ACK.

laanwj commented:
> Not sure what and why you ACK.

Having a single-level SHA256SUMS means that we can put the files in one directory during upload (anyhow, that users are able to verify that without having to move files around or pre-process SHA256SUMS).

achow101 commented:
> As I understand this makes the uploaded binaries have all files in one directory like previous releases?

Unless the binary server validates the SHA256SUMS file, I don't think so.

achow101 commented:
From IRC:

```
<dongcarl> The layout of the SHA256SUMS file is designed such that anyone can download any subtree (as torrent download tools allow you to do), then `cd path/to/all/output/basedir; sha256sum --check --ignore-missing` to verify their files. In that sense it needs to represent represent the tree somewhat accurately.
```

So I guess the directory structure will need to change for the upload. I will add something to `doc/release-process.md` to reflect this.

achow101 committed:
doc: Mention the flat directory structure for uploads

The uploaded binaries need to match the same flat directory structure of
the SHA256SUMS file in order for torrent downloaders to be able to
verify the download without moving files. Mention this in the release
process doc.

wodry commented on contrib/guix/libexec/build.sh:
```diff
@@ -452,11 +452,13 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
+    # To pass in the filename (stored in "{}") safely, it is provided as the second additional argument to "bash"
+    # So that it is available as "$1". The string "bash" is provided as it will be "$0" which is used for some
+    # error messages and warnings. See https://github.com/koalaman/shellcheck/wiki/SC2156 for further information about this trick.
+    find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$1" | xargs sha256sum' bash {} \; \
```
I would suggest simpler:
```
# Recursively get all of the generated output files, but hash them without directory names.
find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -printf '%f\n' | xargs sha256sum \
```

wodry commented on contrib/guix/libexec/build.sh:
```diff
@@ -452,11 +452,13 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
-    } | xargs realpath --relative-base="$PWD" \
-      | xargs sha256sum \
+    # Get all of the generated output files and hash them without directory names.
+    # Because "find" prints filenames with a preceding "./", we need to pass the filenames into "basename" first.
```
Generally I think it would be good to leave the preceding `./` path, because it is precise.
The verifying works with it on Linux, so the removing is not generally needed.
I guess it shall be removed to be compatible with OSs that have other directory naming conventions, e.g. like Microsoft products?
I would find a comment helpful, for what reason the `./` is needed to be removed.

achow101 replied:
Typically SHA256SUMS do not contain paths to the files. Seeing the `./` may be confusing to users who are not familiar with paths.

wodry replied:
BTW [here](http://www.gnu.org/software/coreutils/manual/coreutils.html#sha2-utilities) (leading to the doc of `md5sum --check` invocation) is the doc of SHA256SUMS file:
_The input to this mode of md5sum is usually the output of a prior, checksum-generating run of ‘md5sum’._

achow101 replied:
Yes, but many people do a visual comparison.

dongcarl commented:
Here's an alternative to 752235f0eb74be016ca2b986d8def39960a73391:

```diff
diff --git a/contrib/guix/guix-attest b/contrib/guix/guix-attest
index 1503c330b2..2e2ac0b411 100755
--- a/contrib/guix/guix-attest
+++ b/contrib/guix/guix-attest
@@ -159,6 +159,18 @@ Hint: You may wish to remove the existing attestations and their signatures by
 EOF
 }
 
+# Give a SHA256SUMS file as stdin that has lines like:
+#     0ba536819b221a91d3d42e978be016aac918f40984754d74058aa0c921cd3ea6  a/b/d/c/d/s/bitcoin-22.0rc2-riscv64-linux-gnu.tar.gz
+#     ...
+#
+# Replace each line's file name with its basename:
+#     0ba536819b221a91d3d42e978be016aac918f40984754d74058aa0c921cd3ea6  bitcoin-22.0rc2-riscv64-linux-gnu.tar.gz
+#     ...
+#
+basenameify_SHA256SUMS() {
+    sed -E 's@(^[[:xdigit:]]{64}[[:space:]]+).+/([^/]+$)@\1\2@'
+}
+
 echo "Attesting to build outputs for version: '${VERSION}'"
 echo ""
 
@@ -174,6 +186,7 @@ mkdir -p "$outsigdir"
         cat "${noncodesigned_fragments[@]}" \
             | sort -u \
             | sort -k2 \
+            | basenameify_SHA256SUMS \
                 > "$temp_noncodesigned"
         if [ -e noncodesigned.SHA256SUMS ]; then
             # The SHA256SUMS already exists, make sure it's exactly what we
@@ -201,6 +214,7 @@ mkdir -p "$outsigdir"
         cat "${sha256sum_fragments[@]}" \
             | sort -u \
             | sort -k2 \
+            | basenameify_SHA256SUMS \
                 > "$temp_all"
         if [ -e all.SHA256SUMS ]; then
             # The SHA256SUMS already exists, make sure it's exactly what we
```

fanquake commented:
@achow101 can you incorporate Carls suggestion here.

achow101 commented:
I don't see the benefit of Carl's approach over what I have implemented already.

wodry commented:
In any case it could be good to use the suggested `basenameify_SHA256SUMS` (or should it be called "basenamify_SHA256SUMS" according to "basenamification", I am no English expert?), which is quite good readable.
That would make it possible to remove rather complex redundant code and it's comments. Not that important, though.

dongcarl commented:
For posterity: some previous thoughts from IRC on why it's good to keep the `output/` dir per-host. And flatten it last-minute: https://gnusha.org/bitcoin-builds/2021-08-12.log

```
11:50 < dongcarl> I think #22654 will work well. What it achieves is basically: 1. Updates release-notes.md to mention that we need to flatten the output hier when uploading, 2. Makes the resulting SHA256SUMS file have filenames that are `basename`-ified (no `dirname`)
11:50 < gribble> https://github.com/bitcoin/bitcoin/issues/22654 | guix: Dont include directory name in SHA256SUMS by achow101 · Pull Request #22654 · bitcoin/bitcoin · GitHub
11:53 < dongcarl> I didn't want to touch the output dir hier for guix-build itself because looking at it it might introduce more problems that will slow down rc's and will mess up: 1. Atomicity of the mv OUTDIR ACTUAL_OUTDIR means there are no half-formed outputs 2. In the future we can more granularly bind-mount output dirs into build containers to ensure that one
11:53 < dongcarl> HOST's build is physically unable to tamper with another HOST's output
11:53 < laanwj> i was wondering about the atomicity-per-platform thing
11:54 < laanwj> agree that that makes flattening last minute a better option
```

achow101 commented:
Is it strongly preferred that this use @dongcarl's solution over the one implemented here?

fanquake commented:
Yes I think so.

achow101 commented:
Changed to use @dongcarl's suggestion

wodry commented on doc/release-process.md:
```diff
@@ -207,16 +207,25 @@ cat "$VERSION"/*/all.SHA256SUMS.asc > SHA256SUMS.asc
 
 - Upload to the bitcoincore.org server (`/var/www/bin/bitcoin-core-${VERSION}`):
-    1. The contents of `./bitcoin/guix-build-${VERSION}/output`, except for
+    1. The contents of each `./bitcoin/guix-build-${VERSION}/output/${HOST}` directory, except for
```
Nit: I personally like stating directory names with a traling `/` when possible like here, so code readers can easily recognize that it is a directory, in this case "each `./bitcoin/guix-build-${VERSION}/output/${HOST}/` directory". In line 226 this is done with `/output/`

achow101 replied:
Done

wodry commented on doc/release-process.md:
```diff
@@ -207,16 +207,25 @@ cat "$VERSION"/*/all.SHA256SUMS.asc > SHA256SUMS.asc
 
 
 - Upload to the bitcoincore.org server (`/var/www/bin/bitcoin-core-${VERSION}`):
```
Nit: `/var/www/bin/bitcoin-core-${VERSION}/`

achow101 replied:
Done

Zero-1729 reviewed (commented):
re-ACK 132cae44f2d031bdaa1e459b92ec89ad585dfc9f

fanquake commented:
Guix builds:
```bash
777a1ad25da672e72356d34638fef917649462307e6a857e0c2f4d78b8d3e363  guix-build-132cae44f2d0/output/aarch64-linux-gnu/SHA256SUMS.part
5f307edda9d8785a7051bdc89e667500ddd932324f6a0b225d605a921236cdf5  guix-build-132cae44f2d0/output/aarch64-linux-gnu/bitcoin-132cae44f2d0-aarch64-linux-gnu-debug.tar.gz
99c0fe0aa4952af02384ebd4b84c490351ad462290e4673af5c16d13929d929a  guix-build-132cae44f2d0/output/aarch64-linux-gnu/bitcoin-132cae44f2d0-aarch64-linux-gnu.tar.gz
d30698cf2358edd729c9cae72270df510d2d811debbac1fd2619ad5aeed8eab3  guix-build-132cae44f2d0/output/arm-linux-gnueabihf/SHA256SUMS.part
2a39d1e11e63a1662d644e16d8002ebc778acaac04ac23572666fb57f15f1d07  guix-build-132cae44f2d0/output/arm-linux-gnueabihf/bitcoin-132cae44f2d0-arm-linux-gnueabihf-debug.tar.gz
807494da57ef85914988500f3533257c5ae09b9351ad742ab317cb4326a72fed  guix-build-132cae44f2d0/output/arm-linux-gnueabihf/bitcoin-132cae44f2d0-arm-linux-gnueabihf.tar.gz
31067da3649769233be6ccf8e0aebf0e09af2d70ae649c42011557b111fd0811  guix-build-132cae44f2d0/output/dist-archive/bitcoin-132cae44f2d0.tar.gz
a9b3aad4050e69ff5bc792c2057934da227e793a7ccd73514031bd8c4b48623c  guix-build-132cae44f2d0/output/powerpc64-linux-gnu/SHA256SUMS.part
23a0874ab176e0574351ff60e8a193196b48eb640b0a6f77188d7ceab5f70a3c  guix-build-132cae44f2d0/output/powerpc64-linux-gnu/bitcoin-132cae44f2d0-powerpc64-linux-gnu-debug.tar.gz
bd0d1ac1ffe179ba1d60259cdf718b24f72490c95656842bd2df0fc8f44e794d  guix-build-132cae44f2d0/output/powerpc64-linux-gnu/bitcoin-132cae44f2d0-powerpc64-linux-gnu.tar.gz
ab4d8038e2c1372d2beaaa5c647ba3b30cfe2e0afb38c2174f1264955dacae25  guix-build-132cae44f2d0/output/powerpc64le-linux-gnu/SHA256SUMS.part
41b493490edf41371fe39d6aaef392bec4f1531e26c746a989729cc1bf54f1a4  guix-build-132cae44f2d0/output/powerpc64le-linux-gnu/bitcoin-132cae44f2d0-powerpc64le-linux-gnu-debug.tar.gz
e6b79baa7ed8410f3d5cb4b2d21278c06ce9c665fd7ffe19c1eb3d2e1c147d91  guix-build-132cae44f2d0/output/powerpc64le-linux-gnu/bitcoin-132cae44f2d0-powerpc64le-linux-gnu.tar.gz
7e5ab13ac196511d14a5cb88272ce377257f0b117e5b69a9d8f64aaab8aea580  guix-build-132cae44f2d0/output/riscv64-linux-gnu/SHA256SUMS.part
155146bca7dbe8b499e8ff22395ea71fd770853e075e63293693e6718b1d15b2  guix-build-132cae44f2d0/output/riscv64-linux-gnu/bitcoin-132cae44f2d0-riscv64-linux-gnu-debug.tar.gz
de943b63789da014f642417800197988350be13bdaec2e6ac4dcd8dea0e26b85  guix-build-132cae44f2d0/output/riscv64-linux-gnu/bitcoin-132cae44f2d0-riscv64-linux-gnu.tar.gz
7c7d8ad3fc2e28071429a7e25d8235cefddc489ec7d8992c54f4db517f09f5a1  guix-build-132cae44f2d0/output/x86_64-apple-darwin18/SHA256SUMS.part
69767250b6f817cdc424498a60dec00db412c08950eb06d3687814d0442cda02  guix-build-132cae44f2d0/output/x86_64-apple-darwin18/bitcoin-132cae44f2d0-osx-unsigned.dmg
ea42729a477276ac26982a7a22eee940aff567e367fe151c355b21c3a1b5d714  guix-build-132cae44f2d0/output/x86_64-apple-darwin18/bitcoin-132cae44f2d0-osx-unsigned.tar.gz
cb6c48623e78065551cf3ff4ce45dc1f48dab6538dc571d055c9e2b58ceb5178  guix-build-132cae44f2d0/output/x86_64-apple-darwin18/bitcoin-132cae44f2d0-osx64.tar.gz
f49ebd853f247dbbb01ac325f160d37684540e9c6dad9d1be53cc1d1f25b4e13  guix-build-132cae44f2d0/output/x86_64-linux-gnu/SHA256SUMS.part
a4b55db9e3f5f36a7f856da7ad6aba63dcc70f4b3a1a67b5c45681f5b9e35d90  guix-build-132cae44f2d0/output/x86_64-linux-gnu/bitcoin-132cae44f2d0-x86_64-linux-gnu-debug.tar.gz
ba6f9ef3e4ce462019a38df151eed104ab9ebe72eb93fb99e7a54a8c3999a7eb  guix-build-132cae44f2d0/output/x86_64-linux-gnu/bitcoin-132cae44f2d0-x86_64-linux-gnu.tar.gz
4dc9d667ceec5cca1c6fddfa1296bd174da1be04cef015eebe067dc302600ca0  guix-build-132cae44f2d0/output/x86_64-w64-mingw32/SHA256SUMS.part
4d0799e48974fd1fd5023f98e488c66b4a90f753b2de6fd7b214c7193256a10d  guix-build-132cae44f2d0/output/x86_64-w64-mingw32/bitcoin-132cae44f2d0-win-unsigned.tar.gz
a98ca150dd426a66a568cd07038e8c360bbe6b78eecf315bab2b146f42da526c  guix-build-132cae44f2d0/output/x86_64-w64-mingw32/bitcoin-132cae44f2d0-win64-debug.zip
a7647e742faebfa344df195775b20478a0e42ee19cb671508acc9d96394ec2b9  guix-build-132cae44f2d0/output/x86_64-w64-mingw32/bitcoin-132cae44f2d0-win64-setup-unsigned.exe
f202165ef8de36ec8a8c503053e6c6e81f1e4c1e1f8cad3abbb68f045a73fb1d  guix-build-132cae44f2d0/output/x86_64-w64-mingw32/bitcoin-132cae44f2d0-win64.zip
```

fanquake reviewed (approved):
ACK 132cae44f2d031bdaa1e459b92ec89ad585dfc9f

hebasto commented:
Backported in #22629.
